Utilize new gateway fee structure

nategraf opened the pull request:
### Description

In https://github.com/celo-org/celo-blockchain/pull/602 the network will use a new gateway fee as described in https://github.com/celo-org/celo-blockchain/issues/541

This PR adds a new `gatewayFee` field to transaction objects, renames `gasFeeRecipient` to `gatewayFeeRecipient` and renames `gasCurrency` to `feeCurrency` and makes the appropriate logic changes to handle this new scheme.

Note: `walletkit` is untouched by this PR to maintain compatibility with deployed networks.

### Tested

The e2e transfer test has been updated to test the new behavior. All other e2e and unit tests are run as is.

### Related issues

- Related https://github.com/celo-org/celo-blockchain/issues/541

### Backwards compatibility

Oh most certainly not.

Commit:
rename gasCurrency to feeCurrency and gasFeeRecipient to gatewayFeeRecipient and provide defaults

Commit:
revert changes to walletkit

Commit:
augment web3 transaction types with gateway fee information

Commit:
adjust transfer test to match new behavior

Commit:
Merge branch 'master' of github.com:celo-org/celo-monorepo into victor/gateway-fee

Commit:
consolidate branch changes to one line

Commit:
Merge remote-tracking branch 'origin/timmoreton/free-tobin-tax' into victor/gateway-fee

asaj commented on packages/celotool/src/e2e-tests/transfer_tests.ts:
```diff
@@ -464,34 +469,37 @@ describe('Transfer tests', function(this: any) {
                       txOptions: {
-                        gasFeeRecipient: FeeRecipientAddress,
+                        gatewayFeeRecipient: FeeRecipientAddress,
+                        gatewayFee: '0x10000',
```
We should probably have tests for most combinations of {fee recipient unset, fee recipient peer etherbase, fee recipient random address } x {gateway fee set, gateway fee set too low, gateway fee unset}

nategraf replied:
I've added cases to cover all of those. Thanks for the suggestion!

Commit:
bump gas cost for alternative currencies to account for extra credit op

asaj commented on packages/celotool/src/e2e-tests/transfer_tests.ts:
```diff
@@ -561,25 +581,19 @@ describe('Transfer tests', function(this: any) {
               transferToken: CeloContract.StableToken,
               feeToken: CeloContract.StableToken,
-              txOptions: {
-                gasFeeRecipient: FeeRecipientAddress,
```
Was this there for a reason? Or was this a mistake

nategraf replied:
It's intentional. My main motivation for this is that when we are in the `full` sync mode we want the gateway fee fields to be empty. I decided to delete that field because it allows the test to cover that the defaults should be assigned appropriately by the `light` and `ultralight` nodes.

asaj commented on packages/celotool/src/e2e-tests/transfer_tests.ts:
```diff
@@ -696,8 +704,7 @@ describe('Transfer tests', function(this: any) {
               txRes = await runTestTransaction(
                 await transferCeloGold(FromAddress, ToAddress, TransferAmount, {
-                  gasCurrency: stableTokenAddress,
-                  gasFeeRecipient: FeeRecipientAddress,
```
ditto, needed or mistake?

nategraf replied:
See above.

asaj commented on packages/celotool/src/e2e-tests/transfer_tests.ts:
```diff
@@ -774,8 +780,7 @@ describe('Transfer tests', function(this: any) {
                 await transferCeloGold(FromAddress, ToAddress, TransferAmount, {
                   gas,
-                  gasCurrency: await kit.registry.addressFor(CeloContract.StableToken),
-                  gasFeeRecipient: FeeRecipientAddress,
```
ditto

nategraf replied:
See above.

asaj commented on packages/contractkit/README.md:
```diff
@@ -146,10 +146,11 @@ const goldTokenAddress = await kit.registry.addressFor(CeloContract.GoldToken)
-- gasCurrency (address of the ERC20 contract to use to pay for gas)
-- gasFeeRecipient (address of the beneficiary for the gas, the full node)
+- feeCurrency (address of the ERC20 contract to use to pay for gas)
+- gatewayFeeRecipient (address of the beneficiary for the gas, the full node)
```
Lines 151 and 152 need changes

asaj commented on packages/contractkit/src/providers/celo-private-keys-subprovider.ts:
```diff
@@ -114,20 +119,29 @@ export class CeloPrivateKeysWalletProvider extends PrivateKeyWalletSubprovider {
+      if (isEmpty(txParams.gatewayFeeRecipient)) {
         // Fail early. The validator nodes will reject a transaction missing
-        // gas fee recipient anyways.
+        // gateway fee recipient anyways.
```
Is this true? Won't they just not pay the gateway fee?

nategraf replied:
No, and I'm not sure if that was ever true. I've opted to remove the check, but let me know if you think I should keep it and just rewrite the comment instead.

asaj commented on packages/docs/developer-resources/contractkit/examples.md:
```diff
@@ -66,10 +66,11 @@ const goldTokenAddress = await kit.registry.addressFor(CeloContract.GoldToken)
-- gasCurrency (address of the ERC20 contract to use to pay for gas)
-- gasFeeRecipient (address of the beneficiary for the gas, the full node)
+- feeCurrency (address of the ERC20 contract to use to pay for gas)
+- gatewayFeeRecipient (address of the beneficiary for the gas, the full node)
```
lines 71 and 72 need updating

nategraf replied:
I've updated this.

asaj commented on packages/protocol/migrationsConfig.js:
```diff
@@ -13,7 +13,7 @@ const DefaultConfig = {
   },
   blockchainParameters: {
-    gasForNonGoldCurrencies: 134000,
+    gasForNonGoldCurrencies: 166000,
```
Why 32000 more?

nategraf replied:
To match the default set as a part of https://github.com/celo-org/celo-blockchain/pull/602. The payment of the gateway fee adds an additional credit operation, which we have been estimating at 32000 gas.

asaj replied:
I don't think this is exactly correct, but only because I don't believe what we've been doing is exactly correct. So not in the scope of this PR

nategraf replied:
Ya I was wondering why we don't actually meter the ERC20 calls and either charge only what is used, or keep the intrinsic gas, then issue a refund if its under. Is there a reason for that?

asaj commented on packages/protocol/contracts/common/FeeCurrencyWhitelist.sol:
```diff
@@ -2,14 +2,14 @@ pragma solidity ^0.5.3;
 import "openzeppelin-solidity/contracts/ownership/Ownable.sol";
 
-import "./interfaces/IGasCurrencyWhitelist.sol";
+import "./interfaces/IFeeCurrencyWhitelist.sol";
```
nit: In places where it isn't obvious (e.g. tx params) should we call this `TxFeeCurrencyWhitelist`?

Don't feel strongly

nategraf replied:
I can see it both ways. I think my slight preference is toward `FeeCurrencyWhitelist`

asaj reviewed (commented):
Mostly lgtm, just some small nits

Commit:
fix mutation bug and local signing test

Commit:
Merge branch 'master' of github.com:celo-org/celo-monorepo into victor/gateway-fee

Commit:
edit documentation

Commit:
test each combination of gateway fee params and update gas

Commit:
Merge branch 'master' into victor/gateway-fee

Commit:
gix the various signing-utils

jmrossy commented on .circleci/config.yml:
```diff
@@ -36,7 +36,7 @@ e2e-defaults: &e2e-defaults
     - image: celohq/circleci
   environment:
-    CELO_BLOCKCHAIN_BRANCH_TO_TEST: master 
+    CELO_BLOCKCHAIN_BRANCH_TO_TEST: victor/gateway-fee
```
TODO revert?

nategraf replied:
My plan is to merge https://github.com/celo-org/celo-blockchain/pull/602, then revert this, then merge this PR. Thanks for making sure I didn't miss this.

jmrossy reviewed (dismissed):
I'm unfamiliar with the Gateway concept. Is it just a rebranding of gas?
RE wallet changes, if walletkit isn't updated too, this change will break transactions in the wallet. Let's make sure we have a rollout plan for this :)

nategraf commented:
Marking `do not merge` until https://github.com/celo-org/celo-blockchain/pull/602 is in

nategraf commented:
@jmrossy I have some context in https://github.com/celo-org/celo-blockchain/issues/541. In short, the gateway fee replaces `GasFeeRecipient` as the mechanism to pay transaction serving full-nodes, decoupling full-node incentives from the gas system.

On the `walletkit` changes, it is my understanding that it can't be updated until we deploy a network with these changes. Is that accurate? Either way I should have reached out earlier, but would definitely appreciate some guidance on how to roll out that change.

Commit:
Merge branch 'master' of github.com:celo-org/celo-monorepo into victor/gateway-fee

Commit:
add time to mitigate flake seen in CI

Observed flake in CI here https://app.circleci.com/jobs/github/celo-org/celo-blockchain/15757

Commit:
add retries to validator test to absorb consensus failures

Commit:
Merge branch 'master' of github.com:celo-org/celo-monorepo into victor/gateway-fee

Commit:
reduce assert precision to avoid test flake

Flake in protocol test observed in CI
https://app.circleci.com/jobs/github/celo-org/celo-monorepo/67095

Commit:
Merge branch 'master' into victor/gateway-fee

asaj commented on packages/celotool/src/e2e-tests/validator_order_tests.ts:
```diff
@@ -42,6 +42,9 @@ describe('governance tests', () => {
 
     it('properly orders validators randomly', async function(this: any) {
       this.timeout(160000 /* 160 seconds */)
+      // If a concensus round fails during this test, the results are inconclusive.
```
nit: spelling

asaj reviewed (approved):
✅

Commit:
Merge branch 'master' into victor/gateway-fee

Commit:
fix my dog-gawful spelling

Commit:
Merge branch 'victor/gateway-fee' of github.com:celo-org/celo-monorepo into victor/gateway-fee

asaj commented:
> I'm unfamiliar with the Gateway concept. Is it just a rebranding of gas?

Not exactly, more details here: https://github.com/celo-org/celo-blockchain/issues/541

asaj commented:
> RE wallet changes, if walletkit isn't updated too, this change will break transactions in the wallet. Let's make sure we have a rollout plan for this :)

Unfortunately, walletkit depends on deployed environments to build (currently it looks like integration). So we will need to deploy to integration to get this to work.

Any idea how much is left to migrate from walletkit to contractkit?

Commit:
Merge branch 'master' into victor/gateway-fee

Commit:
Merge branch 'master' of github.com:celo-org/celo-monorepo into victor/gateway-fee

Commit:
Merge branch 'victor/gateway-fee' of github.com:celo-org/celo-monorepo into victor/gateway-fee

Commit:
Merge branch 'master' of github.com:celo-org/celo-monorepo into victor/gateway-fee

Commit:
Revert "revert changes to walletkit"

This reverts commit f90d2b5f59dcd16354219468d4de87cdd440d90a.

Commit:
remove unused refernce to stale contract

Commit:
fix signing

Commit:
Merge branch 'master' of github.com:celo-org/celo-monorepo into victor/gateway-fee

Commit:
Merge branch 'master' of github.com:celo-org/celo-monorepo into victor/gateway-fee

Commit:
fixup walletkit changes for gateway fee

Commit:
Merge branch 'master' of github.com:celo-org/celo-monorepo into victor/gateway-fee
